helix-core/src/main/java/org/apache/helix/controller/stages/MessageGenerationPhase.java

```diff
@@ -271,6 +271,23 @@ private void generateMessage(final Resource resource, final BaseControllerDataPr
     } // end of for-each-partition
   }
 
+  private boolean shouldCreateSTCancellation(Message pendingMessage, String currentState,
```

Comment:
One question: As participants and controller are async with each other, what would happen when the participant is processing the previous message (pending message is still there) and once the cancellation message is being sent, then ST has been executed by the participant? This would potentially happen when pipeline processing time gets extended.

Reply:
It could happen. Then the ST sent by participant will be completed, before the cancellation arrives at the participant. This is exactly what we have seen: when the partition reset ST is processed really fast by participant, controller cannot even read the pending message, so the partition reset is successful without being cancelled. But the success relies on the timing. So we want to skip the error reset ST to make sure the ST is not cancelled.